onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -680,15 +806,17 @@
 
   const auto& input = node_unit.Inputs()[0].node_arg.Name();
   const auto& output = node_unit.Outputs()[0].node_arg.Name();
-  ORT_RETURN_IF_ERROR(shaper.Identity(input, output));
-  const OperandType output_operand_type(operand_types.at(input).type, shaper[output]);
+  Shape output_shape;
+  ORT_RETURN_IF_ERROR(GetShapeInfoFromNodeArg(model_builder, output, output_shape));
+  const OperandType output_operand_type(operand_types.at(input).type, output_shape);
+  shaper.AddShape(output, output_shape);
 
```

Comment:
What's the reason for adding the shape to Shaper? Can we not just retrieve that same info using GetShapeInfoFromNodeArg if needed in the future? Possibly the internals of Shaper could make that call as needed?

Reply:
> What's the reason for adding the shape to Shaper? Can we not just retrieve that same info using GetShapeInfoFromNodeArg if needed in the future?

Yes. that's right. 

but if I remove these lines for adding the shape to Shaper, it will fail bunch of UT test cases in execution stage with a ` Status Message: unordered_map::at: key not found` error status return.

I investigated a bit and found that, 
take SliceOpBuilder as an example(Slice does not require any shaper inference impl code previously, but it has this single step to add output_shaper into Shaper: 
https://github.com/microsoft/onnxruntime/blob/8483b9c6e3bb391525e5c8249bdc9ce0cc1a5522/onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc#L2728-L2731
if remove this line here for slice, SliceOpTest will fail at https://github.com/microsoft/onnxruntime/blob/8483b9c6e3bb391525e5c8249bdc9ce0cc1a5522/onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc#L375

which is further from https://github.com/microsoft/onnxruntime/blob/8483b9c6e3bb391525e5c8249bdc9ce0cc1a5522/onnxruntime/core/providers/nnapi/nnapi_builtin/model.cc#L64-L68
where we retrieve shape info using name from shaper class. 

so I guess to modify this may involve more changes and possibly may require remove the shaper struct completely in nnapi if I understand correctly. 

There may be some other ways to work around this tho, open to suggestions.

Reply:
I was thinking when you construct Shaper you can give it a GraphViewer instance. Shaper::operator[] can look for a shape in shape_map_ first (for newly created shapes that the EP added) and if not found return the shape from the NodeArg in the GraphViewer for `key`.

That provides the same functionality without having to copy shape info into shape_map_ that is already present in the graph. AddShape would only need to be called for modified/new shapes.

Reply:
yeah this makes sense. thanks for the clarification.

addressing this now, along with some other changes.